xcm/pallet-xcm/src/lib.rs

```diff
@@ -635,87 +1147,133 @@ pub mod pallet {
-											weight,
-											max_weight,
-										);
-										Self::deposit_event(e);
-										return 0
-									}
-									let dispatch_origin = Origin::Response(origin.clone()).into();
-									match call.dispatch(dispatch_origin) {
-										Ok(post_info) => {
-											let e =
-												Event::Notified(query_id, pallet_index, call_index);
-											Self::deposit_event(e);
-											post_info.actual_weight
-										},
-										Err(error_and_info) => {
-											let e = Event::NotifyDispatchError(
-												query_id,
-												pallet_index,
-												call_index,
-											);
-											Self::deposit_event(e);
-											// Not much to do with the result as it is. It's up to the parachain to ensure that the
-											// message makes sense.
-											error_and_info.post_info.actual_weight
-										},
-									}
-									.unwrap_or(weight)
-								} else {
-									let e = Event::NotifyDecodeFailed(
+			match (response, Queries::<T>::get(query_id)) {
+				(
+					Response::Version(v),
+					Some(QueryStatus::VersionNotifier { origin: expected_origin, is_active }),
+				) => {
+					let origin: MultiLocation = match expected_origin.try_into() {
+						Ok(o) if &o == origin => o,
+						Ok(o) => {
+							Self::deposit_event(Event::InvalidResponder(
+								origin.clone(),
+								query_id,
+								Some(o),
+							));
+							return 0
+						},
+						_ => {
+							Self::deposit_event(Event::InvalidResponder(
+								origin.clone(),
+								query_id,
+								None,
+							));
+							// TODO #3735: Correct weight for this.
+							return 0
+						},
+					};
+					// TODO #3735: Check max_weight is correct.
+					if !is_active {
+						Queries::<T>::insert(
+							query_id,
+							QueryStatus::VersionNotifier {
+								origin: origin.clone().into(),
+								is_active: true,
+							},
+						);
+					}
+					// We're being notified of a version change.
+					SupportedVersion::<T>::insert(
+						XCM_VERSION,
+						LatestVersionedMultiLocation(&origin),
+						v,
+					);
+					Self::deposit_event(Event::SupportedVersionChanged(origin, v));
+					0
+				},
+				(response, Some(QueryStatus::Pending { responder, maybe_notify, .. })) => {
+					let responder = match MultiLocation::try_from(responder) {
+						Ok(r) => r,
+						Err(_) => {
+							Self::deposit_event(Event::InvalidResponderVersion(
+								origin.clone(),
+								query_id,
+							));
+							return 0
+						},
+					};
+					if origin != &responder {
+						Self::deposit_event(Event::InvalidResponder(
+							origin.clone(),
+							query_id,
+							Some(responder),
+						));
+						return 0
+					}
+					return match maybe_notify {
+						Some((pallet_index, call_index)) => {
+							// This is a bit horrible, but we happen to know that the `Call` will
+							// be built by `(pallet_index: u8, call_index: u8, QueryId, Response)`.
+							// So we just encode that and then re-encode to a real Call.
```

Comment:
I must have missed this in my previous reviews, but how come we've managed to end up in this state of affairs?

Reply:
Ah, is it because it's impossible to create a new `Call` struct otherwise?

Reply:
yup